zetaclient/chains/ton/observer/inbound.go

```diff
@@ -0,0 +1,260 @@
+package observer
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"slices"
+
+	"cosmossdk.io/math"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+	"github.com/tonkeeper/tongo/ton"
+
+	"github.com/zeta-chain/node/pkg/coin"
+	toncontracts "github.com/zeta-chain/node/pkg/contracts/ton"
+	"github.com/zeta-chain/node/pkg/ticker"
+	"github.com/zeta-chain/node/zetaclient/chains/ton/liteapi"
+	zctx "github.com/zeta-chain/node/zetaclient/context"
+	"github.com/zeta-chain/node/zetaclient/zetacore"
+)
+
+const (
+	// MaxTransactionsPerTick is the maximum number of transactions to process on a ticker
+	MaxTransactionsPerTick = 100
+)
+
+func (ob *Observer) watchInbound(ctx context.Context) error {
+	app, err := zctx.FromContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	var (
+		chainID         = ob.Chain().ChainId
+		initialInterval = ticker.SecondsFromUint64(ob.ChainParams().InboundTicker)
+		sampledLogger   = ob.Logger().Inbound.Sample(&zerolog.BasicSampler{N: 10})
+	)
+
+	ob.Logger().Inbound.Info().Msgf("WatchInbound started for chain %d", chainID)
+
+	task := func(ctx context.Context, t *ticker.Ticker) error {
+		if !app.IsInboundObservationEnabled() {
+			sampledLogger.Info().Msgf("WatchInbound: inbound observation is disabled for chain %d", chainID)
+			return nil
+		}
+
+		if err := ob.observeInbound(ctx); err != nil {
+			ob.Logger().Inbound.Err(err).Msg("WatchInbound: observeInbound error")
+		}
+
+		newInterval := ticker.SecondsFromUint64(ob.ChainParams().InboundTicker)
+		t.SetInterval(newInterval)
+
+		return nil
+	}
+
+	return ticker.Run(
+		ctx,
+		initialInterval,
+		task,
+		ticker.WithStopChan(ob.StopChannel()),
+		ticker.WithLogger(ob.Logger().Inbound, "WatchInbound"),
+	)
+}
+
+func (ob *Observer) observeInbound(ctx context.Context) error {
+	if err := ob.ensureLastScannedTX(ctx); err != nil {
+		return errors.Wrap(err, "unable to ensure last scanned tx")
+	}
+
+	lt, hashBits, err := liteapi.TransactionHashFromString(ob.LastTxScanned())
+	if err != nil {
+		return errors.Wrapf(err, "unable to parse last scanned tx %q", ob.LastTxScanned())
```

Comment:
Do we need to return here , or can we still try processing the other tx's

Reply:
I think we do. 

Let's say we have `tx1, tx2, tx3, ...`. If we fail here on while processing `tx2`, we don't want to proceed further and  save "scanned cursor" in db for `tx3` (skipping `tx2` forever). 

This call might return err only if some manual alteration of DB's data was made.